### DISCUSSION TITLE
'doqment': a PDF viewer for Dark Reader users

### DISCUSSION THREAD

**Discussion by shivaprsd** [darkreader/darkreader#11435] [Show and tell]
Time and again I see Dark Reader users open issues here (#3585 #8954 #9637) regarding PDFs. In Firefox, Dark Reader does not work on PDFs, while in Chrome, there seems to be issues with the inversion.

@Gusted himself had submitted a patch (mozilla/pdf.js#13676) to the PDF.js team to have a better dark mode in the Firefox PDF viewer, but it was abandoned due to disagreements regarding implementation. Gusted was not satisfied with a CSS filter-only solution, which I guess is what DR currently does in Chrome anyway.

As this was the case, I went ahead and made an alternative PDF viewer browser extension from the PDF.js code. It's called [*doqment*](https://github.com/shivaprsd/doqment) and IMHO it's the closest we can get, for PDFs, to Dark Reader's functionality for the web. Initially it was just a toy, but now I believe it has matured and is nearing production quality. So I present it to the dark-reading community. I welcome suggestions and feedbacks from everyone as to how to improve it further.

Here is a [demo](https://shivaprsd.github.io/doqment/src/pdfjs/web/viewer.html).

@alexanderby @Gusted Will you be willing to try out *doqment* and, if you find it good, recommend it to your users? You have a good deal of experience developing a great eye-care software. So I invite your contributions. I believe *doqment* shares the same vision (pun intended) as Dark Reader. They can work together to cover all use cases on the web.

Regards,
Shiva

**Comment by bershanskiy** [comment]
Hi, @shivaprsd this looks interesting. As far as I understand, doqment is actually supposed to replace the browser's PDF viewer with a slightly patched copy, correct? I have a few suggestions if you are interested:

 - looks fairly advanced, even if it is just a prototype. Is there tentative iteration plan?
 - I'm surprised to see some of the limitations you list. I think some of these can be actually addressed/worked around.
 - your extension requires very broad and very powerful permissions (MV2 version uses `"webRequestBlocking"` with "<all_urls>", literally the most powerful permission there is, resulting in "Access your data for all websites" warning). MV3 version is much better in this regard (optional permission). I spent a fair amount of time making DR compatible with optional permissions, but it is MV3 only for now. I plan to backport this to MV2 eventually. Would you be interested in sharing experience on "permission-less development"?
 - would you be interested in integrating with DR via cross-extension messaging? For example, DR has a setting for ignoring PDF files, turning on and off on some pages/at certain time, color presets, different color modes. Assuming each user explicitly enables this data sharing, two extensions could provide a more seamless integration. In that case explicit endorsement would make a lot of sense.

**Reply by shivaprsd** [reply]
Hi @bershanskiy, thanks for having a look.
> doqment is actually supposed to replace the browser's PDF viewer with a slightly patched copy, correct?

On Firefox, yes, since it makes no sense to have two nearly identical viewers. OTOH on Chrome, I'd like it to be optional.

Regarding your suggestions:
- I intend to have most of the essential features in place by v1.0. Then it would be mostly keeping up with the release cycle of PDF.js and adding any extra features in between. Right now I'm only able to update PDF.js (their) every third minor release or so.
- please, shoot your suggestions/ideas/PRs! Some of them bug me at times.
- the MV2 variant was based on PDF.js' own (shelved) Chrome extension, which was designed as a total replacement for the native viewer. It needs the `<all_url>` permission since PDFs are ideally detected by the response MIME type, so a `*://*/*.pdf`-like URL filter won't work. But then, `webRequestBlocking` was removed from Chrome's MV3, so it couldn't be done. That was when I decided upon a fresh design for MV3 with contextMenus/browserAction based usage, which needs only `activeTab`. This also made sense on Chrome since it allows people to use both viewers as needed (like we keep two media players or two browsers). Later I made the host permission optional to make it essentially "permission-less" at install time. Sure, it needs a bit more coding, but it gives users finer control. It seems it has also made the store submission & review process faster and hassle-free.
- I suppose that could be done. ~On a basic level I only want DR to ignore *doqment*'s viewer HTML page~ (edit: nothing to be done, extension pages are anyway protected). On Firefox, DR could completely ignore PDFs if it could detect *doqment* is installed and enabled. Then, to make it seamless, we could let users import color presets/modes into *doqment* or vice-versa if possible. *doqment* works with fixed colors, grouped into color schemes, currently stored in a static `colors.json` file. Can you explain DR's color model a bit?

**Comment by luo-chuan** [comment]
If you use Firefox, you can edit [userContent.css](https://davidwalsh.name/firefox-user-stylesheet) like this:
```css
@media (prefers-color-scheme: dark){
  .pdfViewer{filter: invert(1) hue-rotate(180deg);}
}
```
If you use [Chrome](https://chromewebstore.google.com/detail/pdf-reader/ieepebpjnkhaiioojkepfniodjmjjihl) or [Edge](https://microsoftedge.microsoft.com/addons/detail/pdf-reader/nhppiemcomgngbgdeffdgkhnkjlgpcdi), install the extention "PDF Reader", then edit the extention option like this:
```css
html[data-theme*="dark"] .pdfViewer{
  filter: invert(1) hue-rotate(180deg);
}
```

**Reply by shivaprsd** [reply]
@luo-chuan :
> Gusted himself had submitted a patch to the PDF.js team to have **a better dark mode** in the Firefox PDF viewer [...] Gusted was not satisfied with a **CSS filter-only solution**, [...]

And, if you would anyway install another extension on Chrome or Edge, why not try [*doqment*](https://chrome.google.com/webstore/detail/ajmcpdllaaklaocodbnllhkaflncmlog), which IMHO does it the better way? 😕

**Comment by luo-chuan** [comment]
> And, if you would anyway install another extension on Chrome or Edge, why not try [_doqment_](https://chrome.google.com/webstore/detail/ajmcpdllaaklaocodbnllhkaflncmlog), which IMHO does it the better way? 😕

I tried it, but I want more freedom of customizing, so the PDF Reader is better for me.
For me, I need not only inverting the color, but also ratoting the hue 180 degree.

**Reply by shivaprsd** [reply]
@luo-chuan Thanks for the feedback! I will try to improve the next release.

**Reply by shivaprsd** [reply]
As I promised, I added the option to customise the CSS used for inverting in the latest update. Users can configure it in extension options:

<img width="403" alt="Screenshot 2024-08-04 at 14 37 34" src="https://github.com/user-attachments/assets/b30ff9de-d025-4d73-a193-59ba46c36ef9">

&nbsp;
Once again thanks for suggesting this feature! @luo-chuan
